### DISCUSSION TITLE
An Error has occurred (bad request) when using postman

### DISCUSSION THREAD

**Discussion by NPSconsultingLLC** [timdorr/tesla-api#286] [Q&A] [answered]
I must be missing something obvious, but I have not figured out what it its. I created my code challenge using swift: 

```
import UIKit
import CryptoKit
import CommonCrypto

let myRandomString = randomString(length: 86)
let hashString = myRandomString.hash256()
let base64EncodedString = hashString.toBase64()

func randomString(length: Int) -> String {
  let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
  return String((0..<length).map{ _ in letters.randomElement()! })
}

extension String {
    func hash256() -> String {
        let inputData = Data(utf8)
        
        if #available(iOS 13.0, *) {
            let hashed = SHA256.hash(data: inputData)
            return hashed.compactMap { String(format: "%02x", $0) }.joined()
        } else {
            var digest = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
            inputData.withUnsafeBytes { bytes in
                _ = CC_SHA256(bytes.baseAddress, UInt32(inputData.count), &digest)
            }
            return digest.makeIterator().compactMap { String(format: "%02x", $0) }.joined()
        }
    }
    func fromBase64() -> String? {
        guard let data = Data(base64Encoded: self) else {
            return nil
        }
        return String(data: data, encoding: .utf8)
    }
    func toBase64() -> String {
        return Data(self.utf8).base64EncodedString()
    }
}
```

Which produced a challenge String: **"OWNmMjc5MGQzY2M1NGUyZTRmOThkZGVhZjliZTNiNzVkM2YyMGZiZmUyYjg2NGE1ZmY0MTQyZDE0NjM5N2Q0Ng=="** 

Which I used to create my postman request with the following headers: 

`client_id:ownerapi
code_challenge:OWNmMjc5MGQzY2M1NGUyZTRmOThkZGVhZjliZTNiNzVkM2YyMGZiZmUyYjg2NGE1ZmY0MTQyZDE0NjM5N2Q0Ng==
code_challenge_method:S256
redirect_uri:https://auth.tesla.com/void/callback
response_type:code
scope:openid email offline_access
state:**20 RANDOM Letters and Numbers**`

When I send this requests with Postman to the URL *https://auth.tesla.com/oauth2/v3/authorize*

I receive a 400 error (Bad request)

Is there something I am misunderstanding regarding creating my challenge string? or the way I'm formatting my request? 

TIA for the assistance.

**Comment by BenjiFTL** [comment, marked as answer]
Hi, I'm also trying to use postman to try and navigate this new landscape. It's a good visual middleman.

I got step 1 to work. Try removing the padding "==" from your code_challenge and see if that does the trick. Otherwise I'd have to see your postman setup

**Reply by BenjiFTL** [reply]
I'm currently stuck on step 2. My callback URL keeps coming back as https://auth.tesla.com/oauth2/v3/authorize from step one, instead of the redirect_uri. Let me know if you get step two working!

**Reply by BenjiFTL** [reply]
Also for step two make sure you have auto redirect turned off in the settings or you wont be able to catch the callback url from the 302 response.

**Reply by NPSconsultingLLC** [reply]
I have tried removing the "=="'s at the end but I'm still getting a bad request. I have also tried making the request directly from the ShortCuts app in iOS with the same result. 

In addition to the headers I added Postman added its own headers here: 
<img width="1193" alt="Screen Shot 2021-02-04 at 10 55 18 AM" src="https://user-images.githubusercontent.com/6217367/106927270-9d6b7880-66d7-11eb-8613-d8da533ea547.png">

I did turn the redirect setting off, is there other items I need to configure with postman? I'm pretty new to the tool, so I could easily be missing some obvious setting, in other words you're not gonna hurt my feelings if there's stupid obvious that I should be trying :)

**Reply by BenjiFTL** [reply]
I just noticed you are encoding for base64 not base64URL, there is a difference. However, I still don't think that's your issue. Make sure for step one you are sending your parameters in the query parameters tab to get to correct URL.

I have step one working, take a look at my discussion thread #277  to see if you're missing anything.

**Reply by BenjiFTL** [reply]
Also do be aware that postman automatically turns on the cookie jar, you might want to turn it off to properly keep track of the cookies manually if you're going to be making multiple requests. If you don't do this it will auto add the cookie in the header tab.

**Reply by NPSconsultingLLC** [reply]
Well damn - **Make sure for step one you are sending your parameters in the query parameters tab to get to correct URL.**  that was the the problem, I was sending them as headers not query parameters 

Thanks for the help !

**Reply by BenjiFTL** [reply]
Glad it helped, hopefully you don't get stuck on step 2 like me.
